tensorflow-tools/src/main/java/org/tensorflow/tools/Shape.java

```diff
@@ -88,6 +137,38 @@ public boolean hasUnknownDimension() {
     return false;
   }
 
+  /**
+   * Returns whether this shape represents a scalar.
+   * @return whether this shape represents a scalar.
+   */
+  public boolean isScalar() {
+    return dimensionSizes != null && dimensionSizes.length == 0;
+  }
+
+  /**
+   * Returns whether this shape is the shape of a vector.
+   * @return whether this shape is the shape of a vector.
+   */
+  public boolean isVector() {
+    return dimensionSizes != null && dimensionSizes.length == 1;
+  }
+
+  /**
+   * Returns whether this shape is the shape of a matrix
+   * @return whether this shape is the shape of a matrix
+   */
+  public boolean isMatrix() {
+    return dimensionSizes != null && dimensionSizes.length == 2;
+  }
+
+  /**
+   * Returns whether the number of dimensions of this shape is unknown.
+   * @return whether the number of dimensions of this shape is unknown.
+   */
+  public boolean isUnknown() {
+    return dimensionSizes == null;
+  }
+
   public long[] asArray() {
```

Comment:
maybe add doc to this one as well? :) 

Might be worth mentioning that changing the values in the returned array will affect the actual shape (or even better, change the code to return a copy of the array, it might be less performant but it is safer).

Reply:
When I had a look at this, I thought about changing it to return a defensive copy, as the way it was seem like an oversight. But it was used in many places, so I did not feel I confident I knew enough about the inner workings to make such a change. This is why I didn't comment it: I wasn't sure about the intention. As per your suggestion, I changed the method to return a defensive copy and added the appropriate comment & test.